man/man1/soil2cdf.1
.TH soil2cdf 1  "USER COMMANDS"
.SH NAME
soil2cdf \- convert binary files into NetCDF files
.SH SYNOPSIS
.B soil2cdf
[\-global] [\-cellsize size] [\-compress level] [[\-attr name=value] ...] [\-descr d] [\-netcdf4]
.I varname gridfile soilfile netcdffile
.SH DESCRIPTION
Program converts binary input files into NetCDF file.
.SH OPTIONS
.TP
\-global
Use global grid for NetCDF file.
.TP
\-cellsize size
Set cell size in grid file to size.
.TP
\-compress level
Compression level for NetCDF4 files. Valid levels are 0 (no compression) to 9.
.TP
\-attr name=value
Set global attribute name in output files to value.
.TP
\-descr d
Set long name in NetCDF file.
.TP
\-netcdf4
File written is in NetCDF4 format, compress flag is enabled.
.TP
varname
Variable name in NetCDF file.
.TP
.I gridfile
Filename of grid data file.
.TP
.I soilfile
Filename of binary soil data file.
.TP
.I netcdffile
Filename of NetCDF file created.
.SH EXAMPLES
.TP
Create data file for soil:
.B soil2cdf
soilcode grid.bin soil.bin soil.nc
.PP
.SH EXIT STATUS
.B soil2cdf
returns a zero exit status if the NetCDF file is created.
Non zero is returned in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
clm2cdf(1), country2cdf(1), bin2cdf(1), regridlpj(1), lpjml(1), txt2clm(1), cru2clm(1), clm(5)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)

### Changed

- Axis names and missing values for NetCDF output are read from JSON file `netcdf.cjson`.
- Warnings in `addheader` replaced by terminating errors.
- Function `openinputfile()` returns an error if cell size differs in JSON metafile.
- Function `openmetafile()` returns an error if binary file is empty.

### Added

- Environment variables `LPJNOPP` and `LPJPEDANTIC` added. If set to `true`, preprocessor is disabled and pedantic checking is enabled, respectively.
- Info about `--help` option added in case of error in command line arguments.
- Option `-config` added to `bin2cdf`, `country2cdf` and `clm2cdf` in order to read axis names and missing values from a JSON file. File `netcdf.json` added as a template.
- Option `-json` added to `arr2clm`, `asc2clm`, `txt2grid`, `txt2clm`, `getcountry`, and `cdf2coord` in order to create additional JSON metafiles.
- Option `-inpath` added to `regridlpj` in order to set directory where data can be found.
- Option `-double` added to `txt2grid` in order to create grid files with coordinate values of type double.
- Option `-timestep` added to `addheader`.
- Check for correct number of time steps added in `openclimate_netcdf.c`.
- Option `sum`, `tsum` and `tmean` to `mathclm` added.


## [5.9.15] - 2024-09-30

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Sibyll Schaphoff (sibylls@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de)

### Added

- Option `-scale` added to `bin2cdf` in order to scale output.
- Option `-notime` added to `bin2cdf` in order to omit time axis.
- `time_bnds`, `lat_bnds`, and `lon_bnds` arrays added to NetCDF output for lpjml runs and the `bin2cdf` utility.
- Check for identical output filenames added in `fscanoutput.c`. 

### Changed

- If `"with_days"` is set to true in the lpjml configuration file, time axis of yearly output is also set in units of days instead of years.
- Time in NetCDF output is always set in the middle between the lower and upper time boundaries.

### Removed

- Unused file `cflux_sum.c` removed.
- Conversion of obsolete file `manage.par` removed from utility `manage2js`.

### Fixed

- Separate output for each year by setting `"name" : "filename_%d"` is now written correctly.
- Number of bands set to 1 in metafile of grid NetCDF output.
- Name of longitude/latitude variables correctly derived in `getlatlon_netcdf.c`.
- Misspelled option `-netdcdf4` corrected to `-netcdf4` in `clm2cdf.c`.
- `lpj_climber4.c` updated to compile without errors.
- Writing different variables into one NetCDF file fixed in `create_netcdf.c`. Only NetCDF outputs with one band can be written to one file.


## [5.9.14] - 2024-09-30

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de), Stephen Wirth (wirth@pik-potsdam.de)
- code review: Sibyll Schaphoff (sibylls@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de)

### Fixed

- `pft->vscal` not set to `NDAYEAR` in `update_annual.c`. (issue #364).
- `freezefrac2soil.c` corrected to avoid division by zero.
- size of option array for `sowing_date_option` and `crop_phu_option` now correctly specified in `fscanconfig.c`.


## [5.9.13] - 2024-09-26

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Sebastian Ostberg (ostberg@pik-potsdam.de), Marie Hemmen (hemmen@pik-potsdam.de)

### Changed

- `USE_NETCDF4` compile option replaced by `"netcdf4"` boolean flag in the LPJmL configuration file. If set compression of NetCDF4 files can be enabled and PFT names are written as strings instead of character arrays.
- Formatting of man pages harmonized.
- If lpjml is compiled without `-DUSE_NETCDF` flag then `fscanconfig.c` and `fscanoutput.c` return with an error for NetCDF files specified for input/output.

### Added

- Option `-netcdf4` added to `bin2cdf`, `clm2cdf`, `country2cdf`, and `drainage2cdf` utility to enable NetCDF4 format.
- Boolean flag `"netcdf4"` added to LPJmL configuration file in order to enable NetCDF4 format.
- Option `-compress` added to `drainage2cdf` utility to enable file compression.
- Man page for `drainage2cdf` added.
- Check for valid compression value added for lpjml and utilities.
- Missing option `"global_netcdf"` added to LPJmL config file.

### Removed

- Obsolete man page for `writeregioncode()` removed.

### Fixed

- Missing deallocation of memory added in `cpl_init.c`, `fscanlandcovermap.c`, `fscanoutput.c`, `writearea.c`, `create_pft_netcdf.c`, `newgrid.c`, `fscanagtreemap.c`, `celldata.c` in case of error.
- Syntax error fixed in `update_daily.c` in IMAGE coupling.
- Typo in `README` corrected.
- `establishmentpft.c`, `cdf2coord.c` and `cdf2grid.c` modified to compile without warnings using gcc.
- Debug flag in `send_token_coupler.c` and `openoutput_coupler.c` corrected to `DEBUG_COUPLER`.
- Code changed to compile without errors/warnings for `-DIMAGE -DCOUPLED` setting.
- Calculation of offsets and counts corrected in `readintdata_netdf.c` and missing loop over bands added.


## [5.9.12] - 2024-09-24

### Contributors

- author: Christoph Müller (cmueller@pik-potsdam.de)
- code review: Werner von Bloh (bloh@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de)

### Added


src/utils/soil2cdf.c
/**************************************************************************************/
/**                                                                                \n**/
/**                s  o  i  l  2  c  d  f  .  c                                    \n**/
/**                                                                                \n**/
/**     Converts soil data file into NetCDF data file                              \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://gitlab.pik-potsdam.de/lpjml                                   \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"

#if defined(USE_NETCDF) || defined(USE_NETCDF4)
#include <netcdf.h>
#include <time.h>

#define error(rc) if(rc) {free(lon);free(lat);fprintf(stderr,"ERROR427: Cannot write '%s': %s.\n",filename,nc_strerror(rc)); nc_close(cdf->ncid); free(cdf);return NULL;}

#define MISSING_VALUE 999
#define USAGE "%s [-scale s] [-compress level] [-descr d] name gridfile soilfile netcdffile\n"

typedef struct
{
  const Coord_array *index;
  int ncid;
  int varid;
} Cdf;

static Cdf *create_cdf(const char *filename,
                       const char *cmdline,
                       const char *name,
                       const char *descr,
                       Coord res,
                       int compress,
                       const Coord_array *array)
{
  Cdf *cdf;
  float *lon,*lat;
  short miss=MISSING_VALUE;
  int i,rc,dim[2];
  String s;
  time_t t;
  int lat_var_id,lon_var_id,lat_dim_id,lon_dim_id;
  cdf=new(Cdf);
  lon=newvec(float,array->nlon);
  if(lon==NULL)
  {
    printallocerr("lon");
    return NULL;
  }
  lat=newvec(float,array->nlat);
  if(lat==NULL)
  {
    printallocerr("lat");
    free(lon);
    free(cdf);
    return NULL;
  }
  cdf->index=array;
  for(i=0;i<array->nlon;i++)
    lon[i]=array->lon_min+i*res.lon;
  for(i=0;i<array->nlat;i++)
    lat[i]=array->lat_min+i*res.lat;
#ifdef USE_NETCDF4
  rc=nc_create(filename,(compress) ? NC_CLOBBER|NC_NETCDF4 : NC_CLOBBER,&cdf->ncid);
#else
  rc=nc_create(filename,NC_CLOBBER,&cdf->ncid);
#endif
  if(rc)
  {
    fprintf(stderr,"ERROR426: Cannot create file '%s': %s.\n",
            filename,nc_strerror(rc));
    free(lon);
    free(lat);
    free(cdf);
    return NULL;
  }
  rc=nc_def_dim(cdf->ncid,LAT_DIM_NAME,array->nlat,&lat_dim_id);
  error(rc);
  rc=nc_def_dim(cdf->ncid,LON_DIM_NAME,array->nlon,&lon_dim_id);
  error(rc);
  rc=nc_put_att_text(cdf->ncid,NC_GLOBAL,"source",strlen(cmdline),cmdline);
  error(rc);
  time(&t);
  snprintf(s,STRING_LEN,"Created for user %s on %s at %s",getuser(),gethost(),
           strdate(&t));
  rc=nc_put_att_text(cdf->ncid,NC_GLOBAL,"history",strlen(s),s);
  error(rc);
  rc=nc_def_var(cdf->ncid,LAT_NAME,NC_FLOAT,1,&lat_dim_id,&lat_var_id);
  error(rc);
  rc=nc_def_var(cdf->ncid,LON_NAME,NC_FLOAT,1,&lon_dim_id,&lon_var_id);
  error(rc);
  rc=nc_put_att_text(cdf->ncid,lon_var_id,"units",strlen("degrees_east"),
                     "degrees_east");
  error(rc);
  rc=nc_put_att_text(cdf->ncid, lon_var_id,"long_name",strlen("longitude"),"longitude");
  error(rc);
  rc=nc_put_att_text(cdf->ncid, lon_var_id,"standard_name",strlen("longitude"),"longitude");
  error(rc);
  rc=nc_put_att_text(cdf->ncid, lon_var_id,"axis",strlen("X"),"X");
  error(rc);
  rc=nc_put_att_text(cdf->ncid,lat_var_id,"units",strlen("degrees_north"),
                     "degrees_north");
  error(rc);
  rc=nc_put_att_text(cdf->ncid, lat_var_id,"long_name",strlen("latitude"),"latitude");
  error(rc);
  rc=nc_put_att_text(cdf->ncid, lat_var_id,"standard_name",strlen("latitude"),"latitude");
  error(rc);
  rc=nc_put_att_text(cdf->ncid, lat_var_id,"axis",strlen("Y"),"Y");
  error(rc);
  dim[0]=lat_dim_id;
  dim[1]=lon_dim_id;
  rc=nc_def_var(cdf->ncid,name,NC_SHORT,2,dim,&cdf->varid);
  error(rc);
#ifdef USE_NETCDF4
  if(compress)
  {
    rc=nc_def_var_deflate(cdf->ncid, cdf->varid, 0, 1,compress);
    error(rc);
  }
#endif
  if(descr!=NULL)
  {
    rc=nc_put_att_text(cdf->ncid, cdf->varid,"long_name",strlen(descr),descr);
    error(rc);
  }
  nc_put_att_short(cdf->ncid, cdf->varid,"missing_value",NC_SHORT,1,&miss);
  rc=nc_put_att_short(cdf->ncid, cdf->varid,"_FillValue",NC_SHORT,1,&miss);
  rc=nc_enddef(cdf->ncid);
  error(rc);
  rc=nc_put_var_float(cdf->ncid,lat_var_id,lat);
  error(rc);
  rc=nc_put_var_float(cdf->ncid,lon_var_id,lon);
  error(rc);
  free(lat);
  free(lon);
  return cdf;
} /* of 'create_cdf' */

static Bool write_short_cdf(const Cdf *cdf,const short vec[],int size)
{
  int i,rc;
  short *grid;
  grid=newvec(short,cdf->index->nlon*cdf->index->nlat);
  if(grid==NULL)
  {
    printallocerr("grid");
    return TRUE;
  }
  for(i=0;i<cdf->index->nlon*cdf->index->nlat;i++)
    grid[i]=MISSING_VALUE;
  for(i=0;i<size;i++)
    grid[cdf->index->index[i]]=vec[i];
  rc=nc_put_var_short(cdf->ncid,cdf->varid,grid);
  free(grid);
  if(rc!=NC_NOERR)
  {
    fprintf(stderr,"ERROR431: Cannot write output data: %s.\n",
            nc_strerror(rc));
    return TRUE;
  }
  return FALSE;
} /* of 'write_short_cdf' */

static void close_cdf(Cdf *cdf)
{
  nc_close(cdf->ncid);
  free(cdf);
} /* of 'close_cdf' */

#endif

int main(int argc,char **argv)
{
#if defined(USE_NETCDF) || defined(USE_NETCDF4)
  FILE *file;
  Coordfile coordfile;
  Coord_array *index;
  Coord *grid,res;
  Cdf *cdf;
  Byte *data;
  int i,ngrid,iarg,compress,rc;
  short *f;
  char *descr,*endptr,*cmdline;
  float lon,lat;
  Filename filename;
  descr=NULL;
  compress=0;
  for(iarg=1;iarg<argc;iarg++)
    if(argv[iarg][0]=='-')
    {
      if(!strcmp(argv[iarg],"-descr"))
      {
        if(argc==iarg+1)
        {
          fprintf(stderr,"Missing argument after option '-descr'.\n"
                 USAGE,argv[0]);
          return EXIT_FAILURE;
        }
        descr=argv[++iarg];
      }
      else if(!strcmp(argv[iarg],"-compress"))
      {
        if(argc==iarg+1)
        {
          fprintf(stderr,"Missing argument after option '-compress'.\n"
                  USAGE,argv[0]);
          return EXIT_FAILURE;
        }
        compress=strtol(argv[++iarg],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid number '%s' for option '-compress'.\n",argv[iarg]);
          return EXIT_FAILURE;
        }
      }
      else
      {
        fprintf(stderr,"Invalid option '%s'.\n"
                USAGE,argv[iarg],argv[0]);
        return EXIT_FAILURE;
      }
    }
    else
      break;
  if(argc<iarg+4)
  {
    fprintf(stderr,"Missing arguments.\n"
            USAGE,argv[0]);
    return EXIT_FAILURE;
  }
  filename.fmt=CLM;
  filename.name=argv[iarg+1];
  coordfile=opencoord(&filename,TRUE);
  if(coordfile==NULL)
    return EXIT_FAILURE;
  ngrid=numcoord(coordfile);
  getcellsizecoord(&lon,&lat,coordfile);
  res.lon=lon;
  res.lat=lat;
  grid=newvec(Coord,numcoord(coordfile));
  if(grid==NULL)
  {
    printallocerr("grid"); 
    return EXIT_FAILURE;
  }
  for(i=0;i<numcoord(coordfile);i++)
    readcoord(coordfile,grid+i,&res);
  closecoord(coordfile);
  file=fopen(argv[iarg+2],"rb");
  if(file==NULL)
  {
    fprintf(stderr,"Error opening '%s': %s.\n",argv[iarg+2],strerror(errno));
    return EXIT_FAILURE;
  }
  if(getfilesize(argv[iarg+2])!=ngrid)
  {
    fprintf(stderr,"Number of cells in '%s' is different from %d in '%s'.\n",
            argv[iarg+2],ngrid,argv[iarg+1]);
    fclose(file);
    return EXIT_FAILURE;
  }
  index=createindex(grid,ngrid,res);
  if(index==NULL)
    return EXIT_FAILURE;
  free(grid);
  cmdline=catstrvec(argv,argc); 
  cdf=create_cdf(argv[iarg+3],cmdline,argv[iarg],descr,res,compress,index);
  free(cmdline);
  if(cdf==NULL)
    return EXIT_FAILURE;
  data=newvec(Byte,ngrid);
  if(data==NULL)
  {
    printallocerr("data");
    return EXIT_FAILURE;
  }
  f=newvec(short,ngrid);
  if(f==NULL)
  {
    printallocerr("f");
    return EXIT_FAILURE;
  }
  rc=fread(data,1,ngrid,file);
  if(rc!=ngrid)
  {
    fprintf(stderr,"Error reading soil data, %d items read.\n",rc);
    return EXIT_FAILURE;
  }
  for(i=0;i<ngrid;i++)
    f[i]=data[i];
  rc=write_short_cdf(cdf,f,ngrid); 
  close_cdf(cdf);
  fclose(file);
  free(data);
  return (rc) ? EXIT_FAILURE : EXIT_SUCCESS;
#else
  fprintf(stderr,"ERROR401: NetCDF is not supported in this version of %s.\n",argv[0]);
  return EXIT_FAILURE;
#endif
} /* of 'main' */
